File: software/examples/c/example_callback.c
#include <stdio.h>
#include <inttypes.h>

#include "ip_connection.h"
#include "bricklet_real_time_clock.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change to your UID

// Callback function for date and time callback
void cb_date_time(uint16_t year, uint8_t month, uint8_t day, uint8_t hour, uint8_t minute,
                  uint8_t second, uint8_t centisecond, uint8_t weekday, int64_t timestamp,
                  void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Year: %d\n", year);
	printf("Month: %d\n", month);
	printf("Day: %d\n", day);
	printf("Hour: %d\n", hour);
	printf("Minute: %d\n", minute);
	printf("Second: %d\n", second);
	printf("Centisecond: %d\n", centisecond);
	printf("Weekday: %d\n", weekday);
	printf("Timestamp: %"PRId64"\n", timestamp);
	printf("\n");
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	RealTimeClock rtc;
	real_time_clock_create(&rtc, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register date and time callback to function cb_date_time
	real_time_clock_register_callback(&rtc,
	                                  REAL_TIME_CLOCK_CALLBACK_DATE_TIME,
	                                  (void *)cb_date_time,
	                                  NULL);

	// Set period for date and time callback to 5s (5000ms)
	// Note: The date and time callback is only called every 5 seconds
	//       if the date and time has changed since the last call!
	real_time_clock_set_date_time_callback_period(&rtc, 5000);

	printf("Press key to exit\n");
	getchar();
	real_time_clock_destroy(&rtc);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/php/ExampleCallback.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletRealTimeClock.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletRealTimeClock;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your Real-Time Clock Bricklet

// Callback function for date and time callback
function cb_dateTime($year, $month, $day, $hour, $minute, $second, $centisecond, $weekday,
                     $timestamp)
{
    echo "Year: $year\n";
    echo "Month: $month\n";
    echo "Day: $day\n";
    echo "Hour: $hour\n";
    echo "Minute: $minute\n";
    echo "Second: $second\n";
    echo "Centisecond: $centisecond\n";

    if ($weekday == BrickletRealTimeClock::WEEKDAY_MONDAY) {
        echo "Weekday: Monday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_TUESDAY) {
        echo "Weekday: Tuesday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_WEDNESDAY) {
        echo "Weekday: Wednesday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_THURSDAY) {
        echo "Weekday: Thursday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_FRIDAY) {
        echo "Weekday: Friday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_SATURDAY) {
        echo "Weekday: Saturday\n";
    } elseif ($weekday == BrickletRealTimeClock::WEEKDAY_SUNDAY) {
        echo "Weekday: Sunday\n";
    }

    echo "Timestamp: $timestamp\n";
    echo "\n";
}

$ipcon = new IPConnection(); // Create IP connection
$rtc = new BrickletRealTimeClock(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Register date and time callback to function cb_dateTime
$rtc->registerCallback(BrickletRealTimeClock::CALLBACK_DATE_TIME, 'cb_dateTime');

// Set period for date and time callback to 5s (5000ms)
// Note: The date and time callback is only called every 5 seconds
//       if the date and time has changed since the last call!
$rtc->setDateTimeCallbackPeriod(5000);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
